homeassistant/components/freedompro/light.py

```diff
@@ -0,0 +1,147 @@
+"""Support for Freedompro light."""
+import json
+import math
+
+from pyfreedompro import put_state
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_HS_COLOR,
+    SUPPORT_BRIGHTNESS,
+    SUPPORT_COLOR,
+    LightEntity,
+)
+from homeassistant.const import CONF_API_KEY
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import COORDINATOR, DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro light."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id][COORDINATOR]
+    devices = [
+        Device(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "lightbulb"
+    ]
+
+    async_add_entities(devices, False)
+
+
+class Device(CoordinatorEntity, LightEntity):
+    """Representation of an Freedompro light."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro light."""
+        super().__init__(coordinator)
+        self._hass = hass
+        self._session = aiohttp_client.async_get_clientsession(self._hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._type = device["type"]
+        self._characteristics = device["characteristics"]
+        self._on = False
+        self._brightness = 0
+        self._saturation = 0
+        self._hue = 0
+
+    @property
+    def supported_features(self):
+        """Supported features for lock."""
+        support = 0
+        if "brightness" in self._characteristics:
+            support |= SUPPORT_BRIGHTNESS
+        if "hue" in self._characteristics:
+            support |= SUPPORT_COLOR
+        return support
+
+    @property
+    def is_on(self):
+        """Return the status of the light."""
+        device = next(
```

Comment:
This is fairly inefficient. We should pre-calculate this on update of the entity, so it doesn't get figured out each time the state is queried.

Reply:
Is there an event that allows me to update the data only upon change?

Reply:
Yes, there are 2 possible methods.

1) In the data update coordinator when gathering data.
2) Implement the update method in the entity:

I would say 2 is the preferred way to handle it, example of 2: <https://github.com/home-assistant/core/blob/dev/homeassistant/components/verisure/alarm_control_panel.py#L85-L97>